### ISSUE TITLE
wrapper class

### ISSUE THREAD

**Issue by ejklike** [ejklike/cnmt1.0#3]
""" Onmt NMT Model base class definition """
import torch.nn as nn

def batch_z(z, lengths, max_len=None):
    """
    input z: (batch, zdim)
    output z: (len, batch, zdim)
    """
    max_len = max_len or lengths.max()
    return z.unsqueeze(0).repeat(maxlen, 1, 1)

class MGNMTModel(nn.Module):
    """
    Core trainable object in OpenNMT. Implements a trainable interface
    for a simple, generic encoder + decoder model.

    Args:
      encoder (onmt.encoders.EncoderBase): an encoder object
      decoder (onmt.decoders.DecoderBase): a decoder object
    """

    def __init__(self, q_inf, srclm, tgtlm, src2tgt, tgt2src):
        super(NMTModel, self).__init__()
        self.q_inf = q_inf
        self.srclm = srclm
        self.tgtlm = tgtlm
        self.src2tgt = src2tgt
        self.tgt2src = tgt2src

    def forward(self, src, tgt, src_lengths, tgt_lengths, 
                bptt=False, with_align=False):
        """Forward propagate a `src` and `tgt` pair for training.
        Possible initialized with a beginning decoder state.

        Args:
            src (Tensor): A source sequence passed to encoder.
                typically for inputs this will be a padded `LongTensor`
                of size ``(len, batch, features)``. However, may be an
                image or other generic input depending on encoder.
            tgt (LongTensor): A target sequence passed to decoder.
                Size ``(tgt_len, batch, features)``.
            lengths(LongTensor): The src lengths, pre-padding ``(batch,)``.
            bptt (Boolean): A flag indicating if truncated bptt is set.
                If reset then init_state
            with_align (Boolean): A flag indicating whether output alignment,
                Only valid for transformer decoder.

        Returns:
            (FloatTensor, dict[str, FloatTensor]):

            * decoder output ``(tgt_len, batch, hidden)``
            * dictionary attention dists of ``(tgt_len, batch, src_len)``
        """
        # Q(z|x,y)
        mu, sigma = self.q_inf(src, tgt, src_lengths, tgt_lengths)
        
        # Sample z
        noise = mu.clone().normal_()
        z = mu + noise * sigma

        # create batches of z
        z_src = batch_z(z, src_lengths)
        z_tgt = batch_z(z, tgt_lengths)

        # LM(x), LM(y)
        prob_srclm = self.srclm(z_src, src, src_lengths, src_emb)
        prob_tgtlm = self.tgtlm(z_tgt, tgt, tgt_lengths, tgt_emb)

        

        dec_in_tgt = tgt[:-1]  # exclude last target from inputs

        enc_state, memory_bank, lengths = self.encoder(src, lengths)

        if bptt is False:
            self.decoder.init_state(src, memory_bank, enc_state)
        dec_out, attns = self.decoder(dec_in, memory_bank,
                                      memory_lengths=lengths,
                                      with_align=with_align)
        return dec_out, attns

    def update_dropout(self, dropout):
        self.srclm.update_dropout(dropout)
        self.tgtlm.update_dropout(dropout)
        self.src2tgt.update_dropout(dropout)
        self.tgt2src.update_dropout(dropout)

**Comment by ejklike** [comment]
[ ] add opt.zdim
[ ] add z input for enc & dec
[ ]

**Comment by ejklike** [comment]
import torch
import torch.nn as nn


class LANMTModel(nn.Module):
    """
    Core trainable object in OpenNMT. Implements a trainable interface
    for a simple, generic encoder + decoder model.

    Args:
      encoder (onmt.encoders.EncoderBase): an encoder object
      decoder (onmt.decoders.DecoderBase): a decoder object
    """

    def __init__(self, encoder, prior_estimator,
                 cross_encoder, posterior_estimator, 
                 bottleneck, latent2vec, decoder):
        super(LANMTModel, self).__init__()
        self.encoder = encoder
        self.prior_estimator = prior_estimator
        self.cross_encoder = cross_encoder
        self.posterior_estimator = posterior_estimator
        self.bottleneck = bottleneck
        self.latent2vec = latent2vec
        self.decoder = decoder

    def forward(self, src, tgt, lengths, bptt=False):
        """Forward propagate a `src` and `tgt` pair for training.
        Possible initialized with a beginning decoder state.

        Args:
            src (Tensor): A source sequence passed to encoder.
                typically for inputs this will be a padded `LongTensor`
                of size ``(len, batch, features)``. However, may be an
                image or other generic input depending on encoder.
            tgt (LongTensor): A target sequence passed to decoder.
                Size ``(tgt_len, batch, features)``.
            lengths(LongTensor): The src lengths, pre-padding ``(batch,)``.
            bptt (Boolean): A flag indicating if truncated bptt is set.
                If reset then init_state

        Returns:
            (FloatTensor, dict[str, FloatTensor]):

            * decoder output ``(tgt_len, batch, hidden)``
            * dictionary attention dists of ``(tgt_len, batch, src_len)``
        """

        # share molecular encoder
        src_emb, src_enc, lengths = self.encoder(src, lengths)

        # get prior estimates
        prior_estimates = self.prior_estimator(src_enc)
        prior_estimates = torch.mean(prior_estimates, dim=0, keepdims=True)
        
        # get posterior estimates
        q_enc = self.cross_encoder(src, tgt)
        q_estimates = self.posterior_estimator(q_enc)
        q_estimates = torch.mean(q_estimates, dim=0, keepdims=True)

        # sample from q
        z = self.bottleneck(q_estimates, sampling=True)
        z_vec = self.latent2vec(z)

        if bptt is False:
            self.decoder.init_state(src, src_enc, src_emb)
        dec_in = tgt[:-1]  # exclude last target from inputs
        dec_out, attns = self.decoder(dec_in, src_enc, z_vec,
                                      memory_lengths=lengths)
        return dec_out, attns, prior_estimates, q_estimates

    def predict(self, src, tgt, lengths, sampling=True):
        """Forward propagate a `src` and `tgt` pair for training.
        Possible initialized with a beginning decoder state.

        Args:
            src (Tensor): A source sequence passed to encoder.
                typically for inputs this will be a padded `LongTensor`
                of size ``(len, batch, features)``. However, may be an
                image or other generic input depending on encoder.
            tgt (LongTensor): A target sequence passed to decoder.
                Size ``(tgt_len, batch, features)``.
            lengths(LongTensor): The src lengths, pre-padding ``(batch,)``.

        Returns:
            (FloatTensor, dict[str, FloatTensor]):

            * decoder output ``(tgt_len, batch, hidden)``
            * dictionary attention dists of ``(tgt_len, batch, src_len)``
        """

        # share molecular encoder
        src_emb, src_enc, lengths = self.encoder(src, lengths)

        # get prior estimates
        prior_estimates = self.prior_estimator(src_enc)
        prior_estimates = torch.mean(prior_estimates, dim=0, keepdims=True)

        # sample from q
        z = self.bottleneck(prior_estimates, sampling=sampling)
        z_vec = self.latent2vec(z)

        self.decoder.init_state(src, src_enc, src_emb)
        dec_in = tgt[:-1]  # exclude last target from inputs
        dec_out, attns = self.decoder(dec_in, src_enc, z_vec,
                                      memory_lengths=lengths)
        return dec_out, attns

    def update_dropout(self, dropout):
        self.encoder.update_dropout(dropout)
        self.cross_encoder.update_dropout(dropout)
        self.decoder.update_dropout(dropout)
